### PR TITLE
Core module

### PR DESCRIPTION
**Pull request by lpil**
Related to https://github.com/alpaca-lang/alpaca/issues/232

I can't get this to compile. Seems to crash the compiler no matter what I comment out. :/
Even had a file with all lines except the module declaration crash rebar3. Not sure what's going on here.

### CODE REVIEW

**Comment by lpil** [comment]
Seems we're missing travis configuration? There's none on master for me to use

**Comment by j14159** [comment]
@lpil I'll add Travis config this morning (PST)

**Comment by j14159** [comment]
@lpil travis config added, would you mind rebasing on `master`?  Thanks!

**Comment by lpil** [comment]
Rebased on master @j14159 . Seems to be having the same problem as locally. Would you be able to help me debug? I can't work out why the compiler is crashing.

**Comment by j14159** [comment]
Interesting!  Looks like a break in the scanner.  If you'd like to replicate, I checked this quickly by:
- adding `core.alp` to Alpaca's `test_files` folder
- adding a quick compilation test to `alpaca.erl`, just tries to compile `test_files/core.alp`

`rebar3 eunit` turned up a stack trace, here's the first 3 items:

```[{alpaca_scanner,infer_breaks,1,
                             [{file,
                                  "/home/jeremy/code/erlang/alpaca/_build/test/lib/alpaca/src/alpaca_scanner.erl"},
                              {line,67}]},
                         {alpaca_scanner,scan,1,
                             [{file,
                                  "/home/jeremy/code/erlang/alpaca/_build/test/lib/alpaca/src/alpaca_scanner.erl"},
                              {line,12}]},
                         {alpaca_ast_gen,parse_module,1,
                             [{file,
                                  "/home/jeremy/code/erlang/alpaca/_build/test/lib/alpaca/src/alpaca_ast_gen.erl"},
                              {line,142}]},
```
You might want to try that with `make clean compile eunit` in Alpaca `master` as that will build and use `master` as the compiler.

I can't dig much further today but I'll take a deeper look this weekend unless you beat me to it :)

**Comment by lpil** [comment]
I'm afraid I don't know much about how the compiler works, I'll need to do some reading! Don't wait for me :)

**Comment by j14159** [comment]
No worries, I'll try to get to it over the next few days to a week.  We'll probably need to change to building with Alpaca `master` branch though.

File: src/core.alp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+--     |> (assert.equal 3)
+
+
+{-| Function composition, passing results along in the suggested direction. For
+    example, the following code checks if the square root of a number is odd:
+
+        sqrt >> isEven >> not
+-}
+val (>>) 'a 'b 'c : fn (fn 'a -> 'b) (fn 'b -> 'c) -> (fn 'a -> 'c)
+let (>>) f g x =
```
Comment (j14159) [inline src/core.alp]: Found the problem, `>>` is reserved for binaries (as is `<<`) which is tripping up how the scanner infers breaks between functions (it tracks whether or not it's within a binary expression).

**Comment by j14159** [comment]
@lpil further to my comment above, I'm not sure what the best error to report here is.  Ideally it would be a parse error so that we can report a location in the file but I'm not sure we can defer it that far yet.  I'll think about this a bit more.

File: src/core.alp (hunk 1 above)
Comment (lpil) [reply src/core.alp]: Ah yes! Good catch, thank you. What operator would you like us to use for forwards composition?

Comment (lepoetemaudit) [reply src/core.alp]: I've previously defined these in some code of mine as `>>>` and `<<<`.

Comment (lpil) [reply src/core.alp]: Works for me

**Comment by lpil** [comment]
I've updated the PR, removing `>>>` to add in a later PR.

Seems to be failing on CI though I'm unsure why. Compiles and passes test locally.

**Comment by j14159** [comment]
Are you using v0.2.9 built with v0.2.9 locally or 0.2.8?

**Comment by lpil** [comment]
I'm on 2.8

**Pull request closed without merging**